types/ember__application/index.d.ts

```diff
@@ -119,12 +120,12 @@ export default class Application extends Engine {
  * objects is the responsibility of an "owner", which handled its
  * instantiation and manages its lifetime.
  */
-export function getOwner(object: unknown): unknown;
+export function getOwner(object: unknown): Owner | undefined;
```

Comment:
This feels a little bit like `noUncheckedIndexAccess` to me: from a strict correctness perspective, `| undefined` absolutely belongs here, but it seems like quite a pain in cases where you _know_ you'll get something back.

I wonder if there's anything we could do to make this more ergonomic?

Reply:
Yeah, I had a similar thought. We *could* do it by enumerating the known base classes. It would be a bit fragile, in that we’d have to remember to update it any time a new base class is added, but that’s actually pretty rare. Off the top of my head, the classes we would have to care about:

- anything descended from `EmberObject` (so literally just `EmberObject` would do)
- Glimmer `Component`

That *might* be it today? However, another trick becomes: when I have done `setOwner(this, someOwner)`, how do I tell the system that *that* will be the case. I think we would need a registry or something. 🤔

Reply:
[TS Playground][1] exploring that idea and confirming it works as expected. While we *could* explore a registry approach, I think it’s actually better *not* to. As for why not: the only things Ember (and therefore these types) can *guarantee* have an `Owner` are framework-managed classes. Having the return type always be `| undefined` for cases where the user has done `setOwner(this, somePassedInOwner)` in e.g. a `constructor` forces users to check themselves and be honest when trying to use that later, presumably with the use of an `assert`, which will catch them if they *haven’t* set it up.

[1]: https://www.typescriptlang.org/play?#code/PTAEAEBcGcFoEsDmA7A9gJwKYCh4FsAHDSUAcigE8DNphM8AjTdAfRdQYCtMBjSUgNzYQEGAhQYc+IuhLlIVGnUbM2qAO7Jmg3IWKgAoivQB5LrxIAzdKjxlw9JumAdufHSKhwkaLLplylNS0jqosPLZEWsj8Qp5iPpL++vKKIcZs0MwAbvA8mB5gXuK+Unqy9grByk5sNgCukPDIiDrJFSaazKDWtvahzhpa6DrS+gDiADb4eMwAwpGo0VY2duSI03izzhF6y6PlJADKOXmYPav9xsBZ6Ln5BwGgAEqojcwnd2cXfeQDwA0mi0AW9IMxYLd7gUhMIwLB4bBQAATTCWZrwJpLaCgBGIkDYKrnADSaE0ADF0ABDWbqDAAazmk0p0Cy2IAvIZjGY3CQAD6gKYzeaLZZCFE8JlYHr1ZB8eBLUCITCQTrDAAUdNJyG5FgAXKASUMKdTMLT0AymSyaABKfWq5hi3iS86WGVyhVKlVddBq1x60AyzVDW2ge3oUD8mUotFaJEwkS40Bg6BAxA4hGgfFjCoAb1AmAAHtQ+AAVRQmSygAC+PzWheLkFghLaEuZ2IAshRxnN8wWwcgkdjBVthXtMDFQDmq9hxc7QJNlaA8F25vrO92hPWLGXqBW1Z6w2rl93rdaAHSQVAGACO9Upkx3mArAB4wwA+NXWmGtlmgTufKFe37QdQAA748y0SkSA5SB0HqTABGrGcnUpKUFxIZcwPyNcKCwhDsC3Uty0sfdlUPTDTnyU8LyvW970fF930-GFZ1Q850NAQEPkozB9Ved50Dwzci23YjSK9dUuMEnjqMvG87wfYjX29D8v2wAiRKI3cSIPb01QAIiISAoNQfTZNohSGMsZThgjAMB1RZpMCRVThIbKzxMPKdzPk+ilLDOyo0c2NXOwIA

Reply:
Agreed, I like this approach!

Reply:
there is no ergonomic here, @ember/appplication seems to wraps @glimmer/...:
```ts
getOwner<O extends object = object>(object: object): O | undefined;
```
https://github.com/glimmerjs/glimmer-vm/blob/master/packages/%40glimmer/owner/index.ts#L14

Reply:
this was posted before updates were added. I've updated a comment in other place, more up to date, about removing custom type, method overloads and just use conditional type